compiler/luci/pass/src/FuseGRUPass.test.cpp

```diff
@@ -0,0 +1,418 @@
+
+    _tanh_node->x(_add_node_5);
+
+    _mul_node_3->x(_sub_node);
+    _mul_node_3->y(_tanh_node);
+
+    _add_node_6->x(_mul_node_1);
+    _add_node_6->y(_mul_node_3);
+
+    _reshape_node->shape(_add_node_6);
+
+    _body_output_node_3->from(_reshape_node);
+  }
+};
+
+class FuseGRUTestNegGraph : public TestIOGraph, public GRUGraphlet
+{
+public:
+  FuseGRUTestNegGraph() = default;
+
+  void init(void)
+  {
+    TestIOGraph::init({1}, {1});
+    GRUGraphlet::init(g());
+
+    invalid_less_const_type();
+
+    _while_node->input(0, _time_node);
+    _while_node->input(1, _time_node);
+    _while_node->input(2, _state_node);
+    _while_node->input(3, _hidden_node);
+    _while_node->input(4, input());
+
+    _while_node->cond_graph(_cond_graph.get());
+    _while_node->body_graph(_body_graph.get());
+
+    _while_out_node->input(_while_node);
+    output()->from(_while_out_node);
+
+    // cond graph
+    _less_node->x(_cond_input_node);
+    _less_node->y(_less_const_node);
+    _cond_output_node->from(_less_node);
+
+    // body graph
+    _add_node_1->x(_body_input_node_1);
+    _add_node_2->x(_body_input_node_2);
+
+    _body_output_node_5->from(_add_node_1);
+    _body_output_node_4->from(_add_node_2);
+
+    _gather_node->params(_body_input_node_2);
+    _fc_node_1->input(_body_input_node_4);
+    _fc_node_1->weights(_fc_weight_1);
+    _fc_node_1->bias(_fc_bias_1);
+    _fc_node_2->input(_gather_node);
+    _fc_node_2->weights(_fc_weight_2);
+    _fc_node_2->bias(_fc_bias_2);
+
+    _split_node_1->input(_fc_node_1);
+    _split_node_2->input(_fc_node_2);
+
+    _split_out_node_1->input(_split_node_1);
+    _split_out_node_2->input(_split_node_1);
+    _split_out_node_3->input(_split_node_1);
+
+    _split_out_node_4->input(_split_node_2);
+    _split_out_node_5->input(_split_node_2);
+    _split_out_node_6->input(_split_node_2);
+
+    _add_node_3->x(_split_out_node_1);
+    _add_node_3->y(_split_out_node_4);
+
+    _add_node_4->x(_split_out_node_3);
+    _add_node_4->y(_split_out_node_6);
+
+    _logistic_node_1->x(_add_node_3);
+
+    _mul_node_1->x(_body_input_node_4);
+    _mul_node_1->y(_logistic_node_1);
+
+    _sub_node->y(_logistic_node_1);
+
+    _logistic_node_2->x(_add_node_4);
+
+    _mul_node_2->x(_split_out_node_2);
+    _mul_node_2->y(_logistic_node_2);
+
+    _add_node_5->x(_split_out_node_5);
+    _add_node_5->y(_mul_node_2);
+
+    _tanh_node->x(_add_node_5);
+
+    _mul_node_3->x(_sub_node);
+    _mul_node_3->y(_tanh_node);
+
+    _add_node_6->x(_mul_node_1);
+    _add_node_6->y(_mul_node_3);
+
+    _reshape_node->shape(_add_node_6);
+
+    _body_output_node_3->from(_reshape_node);
+  }
+};
+
+} // namespace
+
+TEST(FuseGRUPassTest, fuse_pattern1)
+{
+  FuseGRUTestGraph1 g;
+  luci::FuseGRUPass pass;
+
+  g.init();
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(FuseGRUPassTest, fuse_NEG)
+{
+  FuseGRUTestNegGraph g;
+  luci::FuseGRUPass pass;
+
+  g.init();
+
+  EXPECT_FALSE(pass.run(g.g()));
```

Comment:
adding a note about what negative would help understanding `FuseGRUTestNegGraph`.
It's a bunch of nodes connected and I can't catch what it's doing.